File: src/main/java/io/jenkins/plugins/jobtag/TagConfigure.java
package io.jenkins.plugins.jobtag;

import edu.umd.cs.findbugs.annotations.NonNull;

import hudson.BulkChange;
import hudson.Extension;
import hudson.XmlFile;
import hudson.model.PersistentDescriptor;
import jenkins.model.GlobalConfiguration;
import jenkins.model.Jenkins;
import net.sf.json.JSONObject;
import org.kohsuke.accmod.Restricted;
import org.kohsuke.accmod.restrictions.NoExternalUse;
import org.kohsuke.stapler.DataBoundSetter;
import org.kohsuke.stapler.StaplerRequest;

import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

@Extension(ordinal = TagConfigure.ORDINAL)
public class TagConfigure extends GlobalConfiguration implements PersistentDescriptor {

    private static final Logger LOGGER = Logger.getLogger(TagConfigure.class.getName());

    private static final long serialVersionUID = 3285310269140845583L;

    @Restricted(NoExternalUse.class)
    public static final int ORDINAL = 200;

    private boolean caseSenstive = false;

    public synchronized boolean isCaseSenstive() {
        return caseSenstive;
    }

    @DataBoundSetter
    public synchronized void setCaseSenstive(boolean caseSenstive) {
        this.caseSenstive = caseSenstive;
    }

    public static @NonNull TagConfigure get() {
        return GlobalConfiguration.all().getInstance(TagConfigure.class);
    }

    @Override
    public boolean configure(StaplerRequest req, JSONObject json) {
        try (BulkChange bc = new BulkChange(this)) {
            req.bindJSON(this, json);
            bc.commit();
        } catch (IOException exception) {
            LOGGER.log(
                    Level.WARNING, "Exception occurred while committing bulkchange operation.", exception);
            return false;
        }

        save();
        return true;
    }

    @Override
    public synchronized void load() {
        XmlFile file = getConfigFile();
        if (!file.exists()) {
            caseSenstive = false;
        } else {
            super.load();
        }
    }

    @Override
    public synchronized void save() {
        if (BulkChange.contains(this))  {
            return;
        }

        try {
            getConfigFile().write(this);
        } catch (IOException e) {
            LOGGER.log(Level.WARNING, "Failed to save " + getConfigFile(), e);
        }
    }

}

File: src/main/java/io/jenkins/plugins/jobtag/TagConfigure.kt
package io.jenkins.plugins.jobtag

import hudson.BulkChange
import hudson.Extension
import hudson.model.PersistentDescriptor
import jenkins.model.GlobalConfiguration
import net.sf.json.JSONObject
import org.kohsuke.accmod.Restricted
import org.kohsuke.accmod.restrictions.NoExternalUse
import org.kohsuke.stapler.DataBoundSetter
import org.kohsuke.stapler.StaplerRequest

import java.io.IOException
import java.util.logging.Level
import java.util.logging.Logger

@Extension(ordinal = TagConfigure.ORDINAL.toDouble())
class TagConfigure : GlobalConfiguration(), PersistentDescriptor {

    @get:Synchronized
    @set:DataBoundSetter
    @set:Synchronized
    var isCaseSenstive = false

    @get:Synchronized
    @set:DataBoundSetter
    @set:Synchronized
    var disableColor = false

    override fun configure(req: StaplerRequest, json: JSONObject?): Boolean {
        try {
            BulkChange(this).use { bc ->
                req.bindJSON(this, json)
                bc.commit()
            }
        } catch (exception: IOException) {
            LOGGER.log(
                    Level.WARNING, "Exception occurred while committing bulkchange operation.", exception)
            return false
        }

        save()
        return true
    }

    @Synchronized
    override fun load() {
        val file = configFile
        if (!file.exists()) {
            isCaseSenstive = false
            disableColor = false
        } else {
            super.load()
        }
    }

    @Synchronized
    override fun save() {
        if (BulkChange.contains(this)) {
            return
        }

        try {
            configFile.write(this)
            if (disableColor) {
                JobTagPublisher.clearColorStack()
            } else {
                JobTagPublisher.loadColorStack()
            }
        } catch (e: IOException) {
            LOGGER.log(Level.WARNING, "Failed to save $configFile", e)
        }

    }

    companion object {

        private val LOGGER = Logger.getLogger(TagConfigure::class.java.name)

        private val serialVersionUID = 3285310269140845583L

        @Restricted(NoExternalUse::class)
        const val ORDINAL = 200

        fun get(): TagConfigure {
            return GlobalConfiguration.all().getInstance(TagConfigure::class.java)
        }
    }

}
